### PR TITLE
Generate `odata.nextLink` and `odata.deltaLink` properties on delta functions response schemas

### LINKED ISSUE #285: delta functions are not generating next and delta links properties
### Assemblies affected

1.1.0, 1.0.11, 1.0.10

### Steps to reproduce

Convert the Microsoft Graph description, and look at delta functions /applications/microsoft.graph.delta()

### Expected result

```yaml
  /applications/microsoft.graph.delta():
    get:
      responses:
        2XX:
          description: Success
          content:
            application/json:
              schema:
                title: Collection of application
                type: object
                properties:
                  '@odata.nextLink':
                      type: string
                      nullable: true
                  '@odata.deltaLink':
                      type: string
                      nullable: true
                  value:
                    type: array
                    items:
                      anyOf:
                        - $ref: '#/components/schemas/microsoft.graph.application'
      x-ms-docs-operation-type: function
      x-ms-pageable:
        nextLinkName: '@odata.nextLink'
        operationName: listMore
```

### Actual result

```yaml
  /applications/microsoft.graph.delta():
    get:
      responses:
        2XX:
          description: Success
          content:
            application/json:
              schema:
                title: Collection of application
                type: object
                properties:
                  value:
                    type: array
                    items:
                      anyOf:
                        - $ref: '#/components/schemas/microsoft.graph.application'
                        - type: object
                          nullable: true
      x-ms-docs-operation-type: function
      x-ms-pageable:
        nextLinkName: '@odata.nextLink'
        operationName: listMore
```

### Additional detail

In fact I'd even expect the delta response to inherit from the countable response and be a reusable model instead of being defined inline multiple times. like we've done with #270 for entity sets and navigation properties.

### PR DESCRIPTION
**Pull request by millicentachieng**
Fixes #285 

This PR 
- Renames setting from `RefBaseCollectionPaginationCountResponse` to `EnableODataAnnotationReferencesForResponses` to make it apply for all the [control information](http://docs.oasis-open.org/odata/odata-json-format/v4.0/cs01/odata-json-format-v4.0-cs01.html#_Toc365464684) in the response i.e. `@odata.nextLink`, `@odata.deltaLink` and `@odata.count` in our case
- Adds `@odata.nextLink` for all actions and functions that return a collection if `EnablePagination = true`
- Creates referenceable schema named `BaseDeltaFunctionResponse` if there's at least one delta function in the model and `EnableODataAnnotationReferencesForResponses = true`
- References `BaseDeltaFunctionResponse` schema for delta functions if `EnableODataAnnotationReferencesForResponses = true`
- Adds  `@odata.nextLink` and ` @odata.deltaLink` properties to responses of delta functions if  `EnableODataAnnotationReferencesForResponses = false`

**`EnableODataAnnotationReferencesForResponses = false`**
image

**`EnableODataAnnotationReferencesForResponses = true`**
image

### CODE REVIEW

File: src/Microsoft.OpenApi.OData.Reader/Edm/EdmOperationExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    internal static class EdmOperationExtensions
+    {
+        /// <summary>
+        /// Checks whether the EDM is a delta function
+        /// </summary>
+        /// <param name="operation">The EDM operation.</param>
+        /// <returns></returns>
+        public static bool IsDeltaFunction(this IEdmOperation operation)
+        {
+            if (operation.IsFunction() && operation.Name == "delta")
```
Comment (irvinesunday) [inline src/Microsoft.OpenApi.OData.Reader/Edm/EdmOperationExtensions.cs:17]:
```suggestion
            return operation.IsFunction()
                && operation.Name.Equals("delta", StringComparison.OrdinalIgnoreCase);
```
Then `using System;`

File: src/Microsoft.OpenApi.OData.Reader/Edm/EdmOperationExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
+﻿using Microsoft.OData.Edm;
```
Comment (irvinesunday) [inline src/Microsoft.OpenApi.OData.Reader/Edm/EdmOperationExtensions.cs:1]: Include license info.

File: src/Microsoft.OpenApi.OData.Reader/Generator/OpenApiResponseGenerator.cs [hunk 3]
```diff
@@ -191,18 +191,58 @@ public static OpenApiResponse CreateOperationResponse(this ODataContext context,
...
+                        };
+                    }
+                    else
+                    {
+                        baseSchema.Properties.Add(ODataConstants.OdataNextLink);
+                        baseSchema.Properties.Add(ODataConstants.OdataDeltaLink);
+                        schema = baseSchema;
+                    }
+                }
+                else if (context.Settings.EnablePagination)
```
Comment (irvinesunday) [inline src/Microsoft.OpenApi.OData.Reader/Generator/OpenApiResponseGenerator.cs:233]: Include support for `EnableCount=true` as well?

File: src/Microsoft.OpenApi.OData.Reader/Edm/EdmOperationExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    /// <summary>
+    /// Extension methods for <see cref="IEdmOperation"/>
+    /// </summary>
+    internal static class EdmOperationExtensions
+    {
+        /// <summary>
+        /// Checks whether the EDM is a delta function
+        /// </summary>
+        /// <param name="operation">The EDM operation.</param>
+        /// <returns></returns>
```
Comment (irvinesunday) [inline src/Microsoft.OpenApi.OData.Reader/Edm/EdmOperationExtensions.cs:14]: Missing `returns` value.

File: src/Microsoft.OpenApi.OData.Reader/Generator/OpenApiResponseGenerator.cs [hunk 5]
```diff
@@ -191,18 +191,58 @@ public static OpenApiResponse CreateOperationResponse(this ODataContext context,
...
+                                        Id = Constants.BaseDeltaFunctionResponse  // @odata.nextLink + @odata.deltaLink
+                                    }
+                                },
+                                baseSchema
+                            }
+                        };
+                    }
+                    else
+                    {
+                        baseSchema.Properties.Add(ODataConstants.OdataNextLink);
```
[the remark this answers is not in the document]
Comment (baywet) [reply src/Microsoft.OpenApi.OData.Reader/Generator/OpenApiResponseGenerator.cs:228]: $count is not supported by delta functions because it wouldn't mean much in this scenario

File: src/Microsoft.OpenApi.OData.Reader/OpenApiConvertSettings.cs [hunk 6]
```diff
@@ -106,9 +106,9 @@ public class OpenApiConvertSettings
         public bool EnableCount { get; set; }
 
         /// <summary>
-        /// Gets/sets a value indicating whether or not to reference @odata.nextLink and @odata.count in responses
+        /// Gets/sets a value indicating whether or not to reference @odata.nextLink, @odata.deltaLink and @odata.count in responses
         /// </summary>
-        public bool RefBaseCollectionPaginationCountResponse { get; set; } = true;
+        public bool EnableODataAnnotationReferencesForResponses { get; set; } = true;
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/OpenApiConvertSettings.cs:111]: while this name is more intelligible, it'd be a breaking change with the last minor (stable) we shipped. Revert this change please.

File: src/Microsoft.OpenApi.OData.Reader/Operation/EdmActionOperationHandler.cs [hunk 7]
```diff
@@ -64,16 +64,6 @@ protected override void SetRequestBody(OpenApiOperation operation)
         protected override void SetExtensions(OpenApiOperation operation)
         {
             operation.Extensions.Add(Constants.xMsDosOperationType, new OpenApiString("action"));
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Operation/EdmActionOperationHandler.cs:66]: shouldn't this method call the base on (from operation) ?

File: src/Microsoft.OpenApi.OData.Reader/Operation/EdmFunctionOperationHandler.cs [hunk 8]
```diff
@@ -42,16 +42,6 @@ protected override void SetBasicInfo(OpenApiOperation operation)
         protected override void SetExtensions(OpenApiOperation operation)
         {
             operation.Extensions.Add(Constants.xMsDosOperationType, new OpenApiString("function"));
-            if (Context.Settings.EnablePagination && EdmOperation.ReturnType?.TypeKind() == EdmTypeKind.Collection)
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Operation/EdmFunctionOperationHandler.cs:45]: shouldn't this method call the base on (from operation) ?

File: test/Microsoft.OpenAPI.OData.Reader.Tests/Generator/OpenApiResponseGeneratorTests.cs [hunk 9]
```diff
@@ -320,5 +320,116 @@ public void CreateResponseForEdmActionWhenErrorResponsesAsDefaultIsSet(string ac
...
+                ""value"": {
+                  ""type"": ""array"",
+                  ""items"": {
+                    ""anyOf"": [
+                      {
+                        ""$ref"": ""#/components/schemas/microsoft.graph.application""
+                      },
+                      {
+                        ""type"": ""object"",
+                        ""nullable"": true
```
Comment (baywet) [inline test/Microsoft.OpenAPI.OData.Reader.Tests/Generator/OpenApiResponseGeneratorTests.cs:373]: shouldn't this be a oneOf?

**Review by baywet — CHANGES_REQUESTED** [review]

File: src/Microsoft.OpenApi.OData.Reader/Microsoft.OpenAPI.OData.Reader.csproj [hunk 10]
```diff
@@ -23,7 +23,8 @@
     <PackageReleaseNotes>
 - Use convert setting to toggle between referencing @odata.count and @odata.nextLink #282
 - Fixes URL Path parameters of type datetime generated as strings with quotes #262
-- Add support for alternate keys #120
+- Generate odata.nextLink and odata.deltaLink properties on delta functions response schemas #285
+- Add support for alternate key parameters#120
     </PackageReleaseNotes>
     <AssemblyName>Microsoft.OpenApi.OData.Reader</AssemblyName>
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Microsoft.OpenAPI.OData.Reader.csproj:29]: once you rebase and handle conflicts + feedback, can you bump to preview 4 please?

File: src/Microsoft.OpenApi.OData.Reader/Operation/EdmActionOperationHandler.cs (hunk 7 above)
Comment (millicentachieng) [reply src/Microsoft.OpenApi.OData.Reader/Operation/EdmActionOperationHandler.cs:66]: Could you please explain this a bit more?

File: src/Microsoft.OpenApi.OData.Reader/OpenApiConvertSettings.cs (hunk 6 above)
Comment (millicentachieng) [reply src/Microsoft.OpenApi.OData.Reader/OpenApiConvertSettings.cs:111]: I renamed this to also cater for `@odata.deltaLink` since the name was very specific to only `@odata.nextLink` and `@odata.count`. I want to think that the challenge here is we want to have backward compatibility with the version we are targeting. I suggest that we have both settings for now to do the exact same thing and mark ` RefBaseCollectionPaginationCountResponse` as deprecated. The setting will then evaluate to `false` if either one is `false`. 

Thoughts?

File: test/Microsoft.OpenAPI.OData.Reader.Tests/Generator/OpenApiResponseGeneratorTests.cs (hunk 9 above)
Comment (millicentachieng) [reply test/Microsoft.OpenAPI.OData.Reader.Tests/Generator/OpenApiResponseGeneratorTests.cs:373]: Yes, ideally. See the discussion we had [here](https://github.com/microsoft/OpenAPI.NET.OData/issues/190#issuecomment-1177693486) on why we are currently using `anyOf`.

Comment (baywet) [reply test/Microsoft.OpenAPI.OData.Reader.Tests/Generator/OpenApiResponseGeneratorTests.cs:373]: right, I forgot, thanks for the reminder, resolving.

File: src/Microsoft.OpenApi.OData.Reader/Operation/EdmActionOperationHandler.cs (hunk 7 above)
Comment (baywet) [reply src/Microsoft.OpenApi.OData.Reader/Operation/EdmActionOperationHandler.cs:66]: EdmActionOperationHandler derives from EdmOperationOperationHandler. They both implement SetExtensions with different aspects (set different things). It feels like the implementation in EdmActionOperationHandler  should also call base.SetExtensions so all the information is present. Unless I'm missing why we wouldn't set the information from the base?

File: src/Microsoft.OpenApi.OData.Reader/OpenApiConvertSettings.cs (hunk 6 above)
Comment (baywet) [reply src/Microsoft.OpenApi.OData.Reader/OpenApiConvertSettings.cs:111]: to paraphrase, `RefBaseCollectionPaginationCountResponse` would be marked as obsolete and be a set of accessors to `EnableODataAnnotationReferencesForResponses`? This way people know the migration path and we can pull the obsolete property on the next major.

File: src/Microsoft.OpenApi.OData.Reader/Operation/EdmActionOperationHandler.cs (hunk 7 above)
Comment (millicentachieng) [reply src/Microsoft.OpenApi.OData.Reader/Operation/EdmActionOperationHandler.cs:66]: We are calling `base.SetExtensions(operation)` on line 67. Is that what you mean?

Comment (baywet) [reply src/Microsoft.OpenApi.OData.Reader/Operation/EdmActionOperationHandler.cs:66]: hum. sorry, I missed it somehow, ignore my confused self.

File: src/Microsoft.OpenApi.OData.Reader/OpenApiConvertSettings.cs (hunk 6 above)
Comment (baywet) [reply src/Microsoft.OpenApi.OData.Reader/OpenApiConvertSettings.cs:111]: perfect, go for it then.

File: src/Microsoft.OpenApi.OData.Reader/Generator/OpenApiResponseGenerator.cs (hunk 3 above)
Comment (baywet) [reply src/Microsoft.OpenApi.OData.Reader/Generator/OpenApiResponseGenerator.cs:233]: also calling out your attention on this one @millicentachieng in case you missed it.

**Review by baywet — CHANGES_REQUESTED** [review]

File: src/Microsoft.OpenApi.OData.Reader/Edm/EdmOperationExtensions.cs (hunk 1 above)
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Edm/EdmOperationExtensions.cs:17]:
```suggestion
            if (operation.IsFunction() && "delta".Equals(operation.Name, StringComparison.OrdinalIgnoreCase))
```

File: src/Microsoft.OpenApi.OData.Reader/Generator/OpenApiSchemaGenerator.cs [hunk 11]
```diff
@@ -129,20 +129,34 @@ internal static class OpenApiSchemaGenerator
...
-                if (context.Settings.EnablePagination)
-                    schemas[Constants.BaseCollectionPaginationCountResponse].Properties.Add(ODataConstants.OdataNextLink);
+                    if (context.Settings.EnableCount)
+                        schemas[Constants.BaseCollectionPaginationCountResponse].Properties.Add(ODataConstants.OdataCount);
+                    if (context.Settings.EnablePagination)
+                        schemas[Constants.BaseCollectionPaginationCountResponse].Properties.Add(ODataConstants.OdataNextLink);
+                }
+
+                // @odata.nextLink + @odata.deltaLink
+                if (context.Model.SchemaElements.OfType<IEdmFunction>().Any(x => x.IsDeltaFunction()))
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Generator/OpenApiSchemaGenerator.cs:150]:
```suggestion
                if (context.Model.SchemaElements.OfType<IEdmFunction>().Any(static x => x.IsDeltaFunction()))
```

**Review by baywet — CHANGES_REQUESTED** [review]

**Review by baywet — APPROVED** [review]

File: src/Microsoft.OpenApi.OData.Reader/Edm/EdmOperationExtensions.cs (hunk 2 above)
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Edm/EdmOperationExtensions.cs:1]:
```suggestion
using System;
﻿using Microsoft.OData.Edm;
```

**Review by baywet — CHANGES_REQUESTED** [review]

**Review by baywet — APPROVED** [review]
